porch/pkg/engine/engine.go

```diff
@@ -40,10 +41,12 @@ var tracer = otel.Tracer("engine")
 
 type CaDEngine interface {
 	OpenRepository(ctx context.Context, repositorySpec *configapi.Repository) (repository.Repository, error)
-	CreatePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, obj *api.PackageRevision) (repository.PackageRevision, error)
-	UpdatePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, oldPackage repository.PackageRevision, old, new *api.PackageRevision) (repository.PackageRevision, error)
+
+	Create(ctx context.Context, repositoryObj *configapi.Repository, obj api.PackageObject) (runtime.Object, error)
```

Comment:
Sorry for the delay, got pulled into something else this morning. 

I'll change the interface back so that we have separate functions here.

Reply:
I spent some time today on this, I think this comment (which I do strongly agree with) and trying to avoid awkward switch statements with a lot of type casting ([your other comment](https://github.com/GoogleContainerTools/kpt/pull/3469#discussion_r955563726)) led me back to something a little bit closer to https://github.com/GoogleContainerTools/kpt/pull/3469/files/07f6cac489b04f32bea97f70e19fd5ef94246215

I've refactored the code in [these files](https://github.com/GoogleContainerTools/kpt/blob/07f6cac489b04f32bea97f70e19fd5ef94246215/porch/pkg/registry/porch/package.go) to avoid duplication where I can, by moving the common things out to `packageCommon`.  I experimented with generics but didn't see a clear way for them to help here. Happy to revisit that there was something there that I didn't see. 

I think me calling it "boilerplate" earlier was not the most accurate. It is similar, but not identical, to the definition of packagerevisions and so I think keeping it in separate files makes sense, with the common functionality moved to `packageCommon`.  If there are other areas of code that could be refactored out, please let me know.